experiments/data/prepare_smiles_dataset.py

```diff
@@ -0,0 +1,156 @@
+
+
+def _get_smile_string(doc):
+    is_valid = Chem.MolFromSmiles(doc)
+    prefix = INVALID_PREFIX if is_valid is None else VALID_PREFIX
+    return f"{prefix} {doc}{EOS_TOKEN}"
+
+
+def concatenate_samples_without_splitting(dataset, tokenizer, max_length):
+    """concatenate samples into batches upto max_length without
+    splitting any of the individual samples between batches"""
+
+    tok_articles = [tokenizer(x)["input_ids"] for x in dataset]
+    tok_articles = [sample for sample in tok_articles if len(sample) <= max_length]
+    tok_articles = list(itertools.chain.from_iterable(tok_articles))
+    eos_token = tokenizer.encode(EOS_TOKEN)[0]
+
+    concatenated_articles = []
+    p0, p1, last_eos = 0, 1, 0
+    while p1 < len(tok_articles):
+        if tok_articles[p1] == eos_token:
+            if (p1 - p0 + 1) < max_length:
+                # keep track of most recent eos index, continue exploring
+                last_eos = p1
+
+            elif (p1 - p0 + 1) == max_length:
+                # collect whole pointer window
+                concatenated_articles.append(tok_articles[p0 : p1 + 1])
+                last_eos = p1
+                p0 = p1 + 1
+                p1 = p0
+            else:
+                # max_length exceeded, collect only up to last eos
+                concatenated_articles.append(tok_articles[p0 : last_eos + 1])
+                p0 = last_eos + 1
+                p1 = p0
+        p1 += 1
+
+    # collect final batch
+    concatenated_articles.append(tok_articles[p0:])
+    return concatenated_articles
+
+
+def pad_batched_data(dataset, tokenizer, max_length):
+    padded_sequences_all = []
+    attention_masks_all = []
+
+    for article in dataset:
+        if len(article) < max_length:
+            article, attention_masks = pad_sequence(
+                article, max_length, tokenizer.pad_token_id
+            )
+        else:
+            attention_masks = [1] * max_length
+        padded_sequences_all.append(article)
+        attention_masks_all.append(attention_masks)
+
+    return {
+        "input_ids": padded_sequences_all,
+        "token_type_ids": [[0] * max_length] * len(padded_sequences_all),
+        "attention_mask": attention_masks_all,
+    }
+
+
+def run(config):
+    random.seed(seed=SEED)
+
+    dataset = datasets.load_dataset(
+        path=DATASET_PATH,
+        data_dir=None,
+        cache_dir=None,
+        download_mode=None,
+    )
+    dataset = dataset[config.data_split][DATA_TYPE]
+    mixed_data = process_docs(dataset)
+    tokenizer = AutoTokenizer.from_pretrained(
+        pretrained_model_name_or_path=config.model_name,
+        model_max_length=config.context_length,
+    )
+    if not tokenizer.pad_token:
+        tokenizer.add_special_tokens({"pad_token": "<|padding|>"})
+
+    batched_data = concatenate_samples_without_splitting(
+        mixed_data, tokenizer, config.context_length
+    )
+    total_tokens = sum([len(batch) for batch in batched_data])
```

Comment:
Does this get # samples instead of # tokens? as it's summing the length of each batch

Reply:
Each batch in batched_data is a list which contained many samples concatenated together up to a maximum length of 2048 tokens. This total tokens is the total tokens in all the batches before any padding is added.